[satel] Initial contribution

druciak opened the pull request:
This PR adds new binding for Satel Integra Alarm System. This initial contribution contains only basic features implemented, it supports partition, zone, output and shutter devices, also ethm-1 and int-rs bridges. Discovery of devices is implemented, bridge things require manual configuration.

Signed-off-by: Krzysztof Goworek <krzysztof.goworek@gmail.com>

<!-- Thanks for contributing to the openHAB project! -->
<!-- Please describe the goal and effect of your PR here. -->
<!-- Pay attention to the below notes and to *the guidelines* for this repository. -->
<!-- Feel free to delete any comment lines in the template (starting with "<!--"). -->

<!-- TITLE -->

<!-- Please provide a PR summary in the *Title* above, according to the following schema: -->
<!-- - If related to one specific addon: Mention the addon shortname in square brackets -->
<!--   e.g. "[exec]", "[netatmo]" or "[tesla]" -->
<!-- - If the PR is work in progress: Add "[WIP]" -->
<!-- - Give a short meaningful description in imperative mood -->
<!--   e.g. "Add support for device XYZ" or "Fix wrongly handled exception" -->
<!--   for a new add-on/binding: "Initial contribution" -->
<!-- Examples: -->
<!-- - "[homematic] Improve communication with weak signal devices" -->
<!-- - "[timemachine][WIP] Initial contribution" -->
<!-- - "Update contribution guidelines on new signing rules" -->

<!-- DESCRIPTION -->

<!-- Please give a few sentences describing the overall goals of the pull request. -->
<!-- Give enough details to make the improvement and changes of the PR understandable  -->
<!-- to both developers and tech-savy users. -->

<!-- Please keep the following in mind: -->
<!-- - What is the classification of the PR, e.g. Bugfix, Improvement, Novel Addition, ... ? -->
<!-- - Did you describe the PRs motivation and goal? -->
<!-- - Did you provide a link to any prior discussion, e.g. an issue or community forum thread? -->
<!-- - Did you describe new features for the end user? -->
<!-- - Did you describe any noteworthy changes in usage for the end user? --> 
<!-- - Was the documentation updated accordingly, e.g. the addon README? -->
<!-- - Does your contribution follow the coding guidelines: -->
<!--   http://docs.openhab.org/developers/development/guidelines.html -->
<!-- - Did you check for any (relevant) issues from the static code analysis: -->
<!--   http://docs.openhab.org/developers/development/bindings.html#static-code-analysis -->
<!-- - Did you sign-off your work: -->
<!--   http://docs.openhab.org/developers/contributing/contributing#sign-your-work -->

konradhuebner commented:
Hi, I tested the binding and observed the following:
* binding throws exception regarding missing support for 192bit AES keys. It should be noted somewhere how to add this to the java runtime used by openhab.
* twice a second I got a log entry "Connecting to ETHM-1 module at..." => "ETHM-1 module connected successfully" => "Closing connection to ETHM-1 module". This happened immediately after adding the ETHM-1 thing
* Even though I don't use encryption and did not provide an encryption key, the connection failed due to the Java exception; ETHM-1 thing was shown as offline
* After deleting the ETHM-1 thing, the log shows exceptions: "Could not update status, because callback is missing". It seems as if the binding still tries to use the module data to connect to the Integra system, even though the thing had been removed.

Next run will be with updated Java security.

Instead of using ETHM-1 as a Thing, I was wondering if it would not make more sense to configure it directly within the binding and expect it to be existent? I know this would remove all INT-RS users, but probably everyone who uses openhab also owns the ETHM-1 module? That would allow direct configuration through PaperUI => configuration => bindings => Satel and after that support auto-discovery of the other things.

Besides that, I am really happy you continue your work on this binding. Great stuff! Thanks!

druciak commented:
Hi Konrad.

> binding throws exception regarding missing support for 192bit AES keys. It should be noted somewhere how to add this to the java runtime used by openhab.

It is noted in the README.md

> twice a second I got a log entry "Connecting to ETHM-1 module at..." => "ETHM-1 module connected successfully" => "Closing connection to ETHM-1 module". This happened immediately after adding the ETHM-1 thing

No surprise since you haven't had support for 192bit keys. The bridge connects, but cannot properly communicate with the module and therefore closes the connection. It happens on each refresh, that's why you have the message showing over and over.

> Even though I don't use encryption and did not provide an encryption key, the connection failed due to the Java exception; ETHM-1 thing was shown as offline

That's something I need to check. It shouldn't use encryption in case you don't provide encrytption key.

> After deleting the ETHM-1 thing, the log shows exceptions: "Could not update status, because callback is missing". It seems as if the binding still tries to use the module data to connect to the Integra system, even though the thing had been removed.

Will check this as well.

> no exception and also removal of ETHM-1 thing works, but without security code it is still never connecting. Simply showing "Communication Error"

Security code is not required, you should be still able to communicate with the module. Do you have any stack trace in the log regarding this error?

> Instead of using ETHM-1 as a Thing, I was wondering if it would not make more sense to configure it directly within the binding and expect it to be existent? I know this would remove all INT-RS users, but probably everyone who uses openhab also owns the ETHM-1 module? 

I don't think this is a good idea. First, you wouldn't be able to communicate with two or more systems. Second you are aware of: INT-RS owners wouldn't be happy, and believe me, these is at least one INT-RS user, who actually helped me testing the binding on INT-RS.

druciak commented:
> Even though I don't use encryption and did not provide an encryption key, the connection failed due to the Java exception; ETHM-1 thing was shown as offline
Could not reproduce this. I disabled encryption in DLOADX, restored original Java policy files (with limited key length) and created ETHM-1 thing and it connected without any problems. After that I removed the thing without problems, added new with encryption key and after that I was indeed receiving exception, but this is normal. Removed the thing again and it has been removed from OH. Are you sure you don't have any file with Satel binding configuration? Or maybe your OH db is broken, please try to start with new, clean installation of OH2 and then try again.

> After deleting the ETHM-1 thing, the log shows exceptions: "Could not update status, because callback is missing". It seems as if the binding still tries to use the module data to connect to the Integra system, even though the thing had been removed.
As described above I could not reproduce also this issue.

konradhuebner commented:
I tried again. Setup: Brand-new installation on a Pi3 based on the latest openhabian setup. I downloaded the Zulu JCE extensions, copied them to the lib/security folder, rebootet. I created a security key and enabled encrypted integration. The log still shows errors regarding missing encryption support for 192bit keys. 
I then removed the ETHM-1 thing, added it again, without encryption, ending up in:

2017-06-13 09:33:10.960 [INFO ] [.satel.internal.protocol.SatelModule] - Unhandled exception occurred in communication loop, disconnecting.
java.lang.IllegalStateException: Could not update status, because callback is missing

I then restarted openhab, and now it connects succesful, but unencrypted. It discovered all my elements, but shows many log entries identical to this:
2017-06-13 09:39:30.618 [ERROR] [nding.satel.command.SatelCommandBase] - Other error


If there is anything I can do to help you with improving stability, let me know.

druciak commented:
@konradhuebner 

> Setup: Brand-new installation on a Pi3 based on the latest openhabian setup. 

What version do you have exactly? I don't posses Pi3, neither RPi actually, but I will try to test it on the same version of OH2 at least.

> I downloaded the Zulu JCE extensions, copied them to the lib/security folder, rebootet. I created a security key and enabled encrypted integration. The log still shows errors regarding missing encryption support for 192bit keys.

I have no idea what Zulu JCE is. Have you read README.md? As I pointed out before, there is a mention about 192bit keys at the beginning. If you are using Oracle Java, there is a link to unlimited policy files. For OpenJDK you should have unlimited keys out of the box. At least on my Debian it works that way, not sure if it applies also to openhabian.

> It discovered all my elements, but shows many log entries identical to this: 
2017-06-13 09:39:30.618 [ERROR] [nding.satel.command.SatelCommandBase] - Other error

This is normal thing during discovery. The binding tries to find your devices (partitions, zones, etc.) iterating over the range for specific mainboard, but for missing stuff (query for unexistent partition or zone), ETHM-1 returns an error. I will try to remove this from the log, since it may confuse people.

druciak commented:
@kubawolanin I made requested changes.

konradhuebner commented:
@druciak I run the Raspbian OS with the openhab installer, provided by the openhab project as openhabian. They mainly decided to remove the Oracle JVM with the Zulu Embedded JVM, which is OpenJDK compatible. You can find it here: https://www.azul.com/products/zulu-embedded/ 
The openhab version I run is the latest you can get via the APT repository

druciak commented:
@konradhuebner Thanks for the explanation about Zulu Embedded, will try to download and test it on my machine. However it will not be an RPi.
Regarding the version of OH, which APT repository do you mean: stable, testing or snapshot? If snapshot every day there is a new version, so I would need version number in that case.

konradhuebner commented:
I only use stable repository. I will try to turn on logging and provide more details soon.

Von meinem iPhone gesendet

Am 20.06.2017 um 08:42 schrieb druciak <notifications@github.com<mailto:notifications@github.com>>:


@konradhuebner<https://github.com/konradhuebner> Thanks for the explanation about Zulu Embedded, will try to download and test it on my machine. However it will not be an RPi.
Regarding the version of OH, which APT repository do you mean: stable, testing or snapshot? If snapshot every day there is a new version, so I would need version number in that case.

-
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/openhab/openhab2-addons/pull/2355#issuecomment-309659488>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ARP0XnlynP1Rj1SV1n9gKkql0X08XjZoks5sF2mdgaJpZM4N126Z>.

druciak commented:
> I downloaded the Zulu JCE extensions, copied them to the lib/security folder, rebootet. I created a security key and enabled encrypted integration. The log still shows errors regarding missing encryption support for 192bit keys.

Are you sure the files landed in the proper place? It should be jre/lib/security. I checked this Zulu JDK with JCE Extension and it worked like a charm.

> After deleting the ETHM-1 thing, the log shows exceptions: "Could not update status, because callback is missing". 

It happened to me once, it looks like some race condition. I need to investigate. Anyway when you have the bridge configured properly, it shouldn't happen.

druciak commented:
> After deleting the ETHM-1 thing, the log shows exceptions: "Could not update status, because callback is missing".
I tried to reproduce this on version 2.1.0.201705192246 without success. I was adding and removing bridge thing, always without any problems and no mentioned exception in the log. It seems that something has changed in OH2 core and fixed this issue.

druciak commented:
@martinvw @kaikreuzer any chance to get this reviewed in 2017...? ;)

martinvw commented:
> @martinvw @kaikreuzer any chance to get this reviewed in 2017...? ;)

You can already start by rebasing on top of the master and running the most recent static analysis and resolve all errors and warnings.

Please also remove all

```java
/**
  * {@inheritDoc}
  */
```

And make sure that all @authors contain both the author name and the "Initial contribution"

martinvw commented:
I just checked there are only a few binding in front of yours, I ordered a new 4G+ router and unlimited data so should be able to restart reviewing soon :-)

druciak commented:
> You can already start by rebasing on top of the master and running the most recent static analysis and resolve all errors and warnings.

I have rebased it several times already, no problem to rebase it once again. All relevant static analysis errors and warnings should be resolved. Will double check.

> And make sure that all @authors contain both the author name and the "Initial contribution"

Will do.

martinvw commented:
> I have rebased it several times already, no problem to rebase it once again. All relevant static analysis errors and warnings should be resolved. Will double check.

I might indeed be mistaken because the version with `{@inheritDoc}` is not yet released

druciak commented:
Well, many core files were just moved from 1.x version without any modification. If it makes troubles in 2.x I will remove all ocurrences.

paweldudakr commented:
I successfully play with satel binding in my home. I think that this binding should receives positive reviews.

druciak commented:
The binding is available on Eclipse Marketplace: [http://marketplace.eclipse.org/content/satel-binding](http://marketplace.eclipse.org/content/satel-binding).
I should have done this months ago... :(

martinvw commented on addons/binding/org.openhab.binding.satel/src/main/java/org/openhab/binding/satel/internal/command/SatelCommandBase.java:
```diff
@@ -0,0 +1,191 @@
+            case 0x11:
+                errorMsg = "Can not arm, but can use force arm";
+                break;
+            case 0x12:
```
Maybe it could make sense to store these in a map instead and look them up from there. You would only need a separate case for success and for codes not being in the map.

druciak replied:
I am not sure it will be better, now it is in one place, so you can easily check what message is assigned to specific error code. But if you insist, I will change it.

martinvw commented on addons/binding/org.openhab.binding.satel/src/main/java/org/openhab/binding/satel/internal/command/SetClockCommand.java:
```diff
@@ -0,0 +1,50 @@
+/**
+ * Copyright (c) 2010-2017 by the respective copyright holders.
```
Please update ;-)

druciak replied:
Done.

martinvw commented on addons/binding/org.openhab.binding.satel/src/main/java/org/openhab/binding/satel/internal/discovery/SatelDeviceDiscoveryService.java:
```diff
@@ -0,0 +1,181 @@
+    private volatile boolean scanStopped;
+
+    public SatelDeviceDiscoveryService(SatelBridgeHandler bridgeHandler, BundleContext bundleContext)
+            throws IllegalArgumentException {
```
This a runtime exception no need and no valuable information by adding this throws it can also throw an `ArrayIndexOutOfBoundsException` and an `NullPointerException`

druciak replied:
So why is it declared in the super class? :) 
Anyway you're right, removed.

martinvw commented on addons/binding/org.openhab.binding.satel/src/main/java/org/openhab/binding/satel/internal/discovery/SatelDeviceDiscoveryService.java:
```diff
@@ -0,0 +1,181 @@
+                } else if (cmd.getState() != SatelCommand.State.FAILED) {
+                    // serious failure, disconnection or so
+                    scanStopped = true;
+                    logger.error("Unexpected failure during scan for {} using {}", deviceType.name(),
```
In [E.4 of the coding guidelines](http://docs.openhab.org/developers/development/guidelines.html#e-logging) we listed what kind of logging level should be used when. Error should only be used:

> `error` logging should only be used to inform the user that something is tremendously wrong in his setup, the system cannot function normally anymore, and there is a need for immediate action. 

Please change this to use the correct logging levels and if possible change the ONLINE/OFFLINE status of thing / bridge.

martinvw replied:
Please check others as well

martinvw commented on addons/binding/org.openhab.binding.satel/src/main/java/org/openhab/binding/satel/internal/protocol/IntRSModule.java:
```diff
@@ -0,0 +1,128 @@
+            logger.info("INT-RS module connected successfuly");
+            return new SerialCommunicationChannel(serialPort);
+        } catch (NoSuchPortException e) {
+            logger.error("Port {} does not exist", this.port);
```
See above about error logging, here I would like to add that the message which log here should go to the detailMessage of the status update.

druciak replied:
Changed the code a bit and now those errors are propagated to the bridge and sent as "detailMessage" when updating bridge status.

martinvw commented on addons/binding/pom.xml:
```diff
@@ -74,6 +74,7 @@
     <module>org.openhab.binding.rfxcom.test</module>
     <module>org.openhab.binding.russound</module>
     <module>org.openhab.binding.samsungtv</module>
+    <module>org.openhab.binding.satel</module>
```
To have the binding being picked up by the distro, you furthermore need to add it to the feature.xml, again at the alphabetically correct position. If you have a dependency on some transport bundle (e.g. upnp, mdns or serial), make sure to add a line for this dependency as well (see the other bindings as an example).

druciak replied:
Good to know. Why is it not included in the archetype?

druciak replied:
Done.

martinvw commented:
Can you please rebase and update the version to 2.3.0-snapshot and update the copyright years, you should be able to so with the command `mvn license:format`

druciak commented on addons/binding/org.openhab.binding.satel/ESH-INF/binding/binding.xml:
```diff
@@ -0,0 +1,9 @@
+	xsi:schemaLocation="http://eclipse.org/smarthome/schemas/binding/v1.0.0 http://eclipse.org/smarthome/schemas/binding-1.0.0.xsd">
+
+	<name>@text/binding.name</name>
+	<description>@text/binding.description</description>
```
Done.

druciak commented on addons/binding/org.openhab.binding.satel/ESH-INF/thing/ethm-1.xml:
```diff
@@ -0,0 +1,46 @@
+
+		<config-description>
+			<parameter name="host" type="text" required="true">
+				<label>@text/ethm1.host.label</label>
```
Done.

druciak commented on addons/binding/org.openhab.binding.satel/ESH-INF/thing/int-rs.xml:
```diff
@@ -0,0 +1,37 @@
+
+		<config-description>
+			<parameter name="port" type="text" required="true">
+				<label>@text/intrs.port.label</label>
```
"serial-port" is more appropriate ;)
Done.

druciak commented on addons/binding/org.openhab.binding.satel/ESH-INF/thing/system.xml:
```diff
@@ -0,0 +1,65 @@
+
+		<channels>
+			<channel typeId="date_time" id="date_time">
+				<label>@text/system.date_time.label</label>
```
Done.

druciak commented on addons/binding/org.openhab.binding.satel/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,30 @@
+Bundle-ClassPath: .
+Import-Package: 
+ com.google.common.collect,
+ gnu.io;resolution:=optional,
```
In theory - yes. There are two kinds of bridges: ethernet and serial. Most of the users have ethernet bridge, so serial library is not used. Not sure if it works that way.

druciak replied:
Anyway I removed optional resolution.

druciak commented on addons/binding/org.openhab.binding.satel/README.md:
```diff
@@ -0,0 +1,315 @@
+
+This binding supports following thing types:
+
+| Thing type | Description |
```
Done.

druciak commented on addons/binding/org.openhab.binding.satel/src/main/java/org/openhab/binding/satel/SatelBindingConstants.java:
```diff
@@ -0,0 +1,59 @@
+
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+
+import com.google.common.collect.ImmutableSet;
```
Done.

druciak commented on addons/binding/org.openhab.binding.satel/src/main/java/org/openhab/binding/satel/handler/Ethm1BridgeHandler.java:
```diff
@@ -0,0 +1,78 @@
+    @Override
+    public Collection<ConfigStatusMessage> getConfigStatus() {
+        // The bridge IP address to be used for checks
+        final String host = (String) getThing().getConfiguration().get(HOST);
```
Done.

druciak commented on addons/binding/org.openhab.binding.satel/src/main/java/org/openhab/binding/satel/handler/SatelBridgeHandler.java:
```diff
@@ -0,0 +1,219 @@
+    private SatelBridgeConfig config;
+    private SatelModule satelModule;
+    private ScheduledFuture<?> pollingJob;
+    private String userCodeOverride = null;
```
This is to emphasize the fact that we do not override user code by default. This field is set by a command and the rest are initialized at the beginning. 
Anyway "null" removed.

druciak commented on addons/binding/org.openhab.binding.satel/src/main/java/org/openhab/binding/satel/handler/SatelBridgeHandler.java:
```diff
@@ -0,0 +1,219 @@
+
+        if (satelModule != null) {
+            if (pollingJob == null || pollingJob.isCancelled()) {
+                Runnable pollingCommand = new Runnable() {
```
Done.

druciak commented on addons/binding/org.openhab.binding.satel/src/main/java/org/openhab/binding/satel/handler/SatelBridgeHandler.java:
```diff
@@ -0,0 +1,219 @@
+                        }
+
+                        // get list of states that have changed
+                        logger.trace("Sending 'get new states' command");
```
Logging is done for all sent command, this is just to mark the starting point of refresh cycle - it is easier to find this specific log entry and analyze all communication from that point.

druciak commented on addons/binding/org.openhab.binding.satel/src/main/java/org/openhab/binding/satel/handler/SatelBridgeHandler.java:
```diff
@@ -0,0 +1,219 @@
+                        satelModule.sendCommand(new NewStatesCommand(satelModule.getIntegraType().hasExtPayload()));
+                    }
+                };
+                pollingJob = scheduler.scheduleAtFixedRate(pollingCommand, 0, config.getRefresh(),
```
Done.

druciak commented on addons/binding/org.openhab.binding.satel/src/main/java/org/openhab/binding/satel/handler/SatelPartitionHandler.java:
```diff
@@ -0,0 +1,99 @@
+                // arm or disarm, depending on command
+            case ARMED:
+            case REALLY_ARMED:
+                return new ControlObjectCommand(
```
Good point, done.

druciak commented on addons/binding/org.openhab.binding.satel/src/main/java/org/openhab/binding/satel/handler/SatelSystemHandler.java:
```diff
@@ -0,0 +1,137 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.collect.ImmutableSet;
```
Done.

druciak commented on addons/binding/org.openhab.binding.satel/src/main/java/org/openhab/binding/satel/handler/SatelSystemHandler.java:
```diff
@@ -0,0 +1,137 @@
+            if (thingConfig.isCommandOnly()) {
+                return;
+            }
+            updateState(getThing().getChannel(CHANNEL_DATE_TIME).getUID(),
```
Done.

druciak commented on addons/binding/org.openhab.binding.satel/src/main/java/org/openhab/binding/satel/handler/SatelSystemHandler.java:
```diff
@@ -0,0 +1,137 @@
+                }
+                if (dateTime != null) {
+                    return new SetClockCommand(dateTime.getCalendar(), bridgeHandler.getUserCode());
+                }
```
Definitely a bug, done.

druciak commented on addons/binding/org.openhab.binding.satel/src/main/java/org/openhab/binding/satel/internal/BundleTextTranslator.java:
```diff
@@ -0,0 +1,65 @@
+ * @author Krzysztof Goworek - Initial contribution
+ *
+ */
+public class BundleTextTranslator implements AutoCloseable {
```
Is there other way to get localized label of a thing?

druciak replied:
I changed the code to get System thing label using ThingTypeRegistry

druciak replied:
It doesn't use ThingTypeRegistry directly, but through BaseThingHandler. Thing type label is to add new thing labeled as it's type, as there should be only one such thing.

druciak commented on addons/binding/org.openhab.binding.satel/src/main/java/org/openhab/binding/satel/internal/SatelHandlerFactory.java:
```diff
@@ -0,0 +1,129 @@
+import org.openhab.binding.satel.internal.discovery.SatelDeviceDiscoveryService;
+import org.osgi.framework.ServiceRegistration;
+
+import com.google.common.collect.ImmutableSet;
```
Done.

druciak commented on addons/binding/org.openhab.binding.satel/src/main/java/org/openhab/binding/satel/internal/SatelHandlerFactory.java:
```diff
@@ -0,0 +1,129 @@
+    @Override
+    public Thing createThing(ThingTypeUID thingTypeUID, Configuration configuration, ThingUID thingUID,
+            ThingUID bridgeUID) {
+        if (thingUID == null) {
```
You tell me, I saw this somewhere in some binding code. Since I am not ESH expert I just did the same. If not needed, will be removed.

druciak replied:
From super class `BaseThingHandlerFactory.createThing`:
```
        if (thingUID == null) {
            thingUID = ThingFactory.generateRandomThingUID(thingTypeUID);
        }
```
So apparently thingUID can be at this point null. Maybe you thought about thingTypeUID, not thingUID? Leaving this code as is.

druciak replied:
It has changed (this PR has almost a year), but the condition is still present: https://github.com/eclipse/smarthome/blob/master/bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandlerFactory.java#L266
Doesn't look similar?

kaikreuzer replied:
Ok, you're right, that looks similar (and had meanwhile been refactored, so it isn't the same as the code above anymore, so I didn't notice).

druciak commented on addons/binding/org.openhab.binding.satel/src/main/java/org/openhab/binding/satel/internal/command/ControlObjectCommand.java:
```diff
@@ -0,0 +1,72 @@
+            final BitSet newStates = this.controlType.getControlledStates();
+            if (newStates != null && !newStates.isEmpty()) {
+                // add delay to give a chance to process sent command
+                REFRESH_TIMER.schedule(new TimerTask() {
```
In this case TimerTask class is required, which cannot be converted to a functional interface.

druciak commented:
Hi @martinvw , I have applied most of your suggestions. I hope now it looks better and we are getting closer to the merge :)

druciak commented:
@martinvw @kaikreuzer what else is blocking this PR being merged? Is there anything I can do to make it happen?

druciak commented:
Hey @martinvw , I understand all of this, I have similar issues. :-) 
However I assume the only thing to do left is click "Merge", so you don't need much time to do so.

kaikreuzer commented on addons/binding/org.openhab.binding.satel/src/main/java/org/openhab/binding/satel/SatelBindingConstants.java:
```diff
@@ -0,0 +1,62 @@
+    public static final String CHANNEL_USER_CODE = "user_code";
+
+    // Thing properties
+    public static final String PROPERTY_REQUIRES_REFRESH = "requiresRefresh";
```
This looks fishy. Are you misusing Thing properties (that are supposed to be static properties of the remote device) here as local variables for your code?

druciak replied:
This was IMO the best way to mark a thing to refresh.

kaikreuzer replied:
Not a valid one, though.
What does "thing to refresh" actually mean? When do you have to do it?

druciak replied:
After a connection is established, things need to be updated - as far as I remeber. A kind of force polling current state when it is not up to date. I will think about different solution if it is fobidden to use properties for such purpose.

kaikreuzer replied:
The moment when you change the Thing status to ONLINE might be a good one to also query the current state of its channels.

kaikreuzer commented on addons/binding/org.openhab.binding.satel/src/main/java/org/openhab/binding/satel/handler/SatelThingHandler.java:
```diff
@@ -0,0 +1,179 @@
+        return bitset;
+    }
+
+    protected boolean requiresRefresh() {
```
Not clear what this method is used for exactly and how it works.
Note that you should never really come across INITIALIZING Thing status during the lifetime of your handler.

druciak replied:
If the code is here, that means it **did** happen. Not sure if still can.

kaikreuzer replied:
If it did happen, than only due to the fact that you miss to set the Thing status in the initialize() method - you need to set it to either UNKNOWN, ONLINE and OFFLINE there.

druciak replied:
Done.

kaikreuzer commented on addons/binding/org.openhab.binding.satel/src/main/java/org/openhab/binding/satel/handler/SatelThingHandler.java:
```diff
@@ -0,0 +1,179 @@
+    }
+
+    protected void updateSwitch(String channelID, boolean switchOn) {
+        ChannelUID channelUID = new ChannelUID(this.getThing().getUID(), channelID);
```
Note that you can call `updateState()` with the channelID instead of the channelUID as well, so no need to construct the UID here.

druciak replied:
This helps optimizing things in case channel UID is present. The method you mention does the same - creates UID. In case it is available it does not make sense to create it again.

kaikreuzer commented on addons/binding/org.openhab.binding.satel/src/main/java/org/openhab/binding/satel/handler/SatelThingHandler.java:
```diff
@@ -0,0 +1,179 @@
+        String channelId = stateType.toString().toLowerCase();
+        Channel channel = getThing().getChannel(channelId);
+        if (channel == null) {
+            logger.warn("Missing channel for {}", stateType);
```
reduce to debug

druciak replied:
Done.

kaikreuzer commented on addons/binding/org.openhab.binding.satel/src/main/java/org/openhab/binding/satel/handler/SatelBridgeHandler.java:
```diff
@@ -0,0 +1,217 @@
+        logger.debug("New command for {}: {}", channelUID, command.toFullString());
+    }
+
+    protected void initialize(final SatelModule satelModule) {
```
Where do you ever set the Thing status to ONLINE or OFFLINE here (depending on whether the connection can be established or not)?

druciak replied:
This is an abstract class, thing status is set in its subclasses.

kaikreuzer replied:
I didn't find it in any of the sub-classes either, that's why I asked.

druciak replied:
Ethm1BridgeHandler.java line 56, IntRSBridgeHandler.java line 55, SatelBridgeHandler.java lines 56, 59

kaikreuzer replied:
All those places only set it to OFFLINE if some configuration is missing.
My question was: "Where do you ever set the Thing status to ONLINE or OFFLINE here (depending on whether the connection can be established or not)?"

druciak replied:
Apparently I don't understand something. For me it is "SatelBridgeHandler.java lines 56, 59", but maybe you're asking about something else?

kaikreuzer replied:
Maybe I do not fully grasp the internals of the binding then. Lines 56-59 seem to be only executed if some "event" comes along. My (maybe too simple) assumption was that those are events that happen if some activity is already in place. If you can confirm that right after `initialize()` has finished, there will always be some event (where is it triggered?), which makes lines 56-59 being executed, then all should be fine indeed.

druciak replied:
That's right. There's an event with current connection status. Once the connection is established and the module sends a response the new "connected" status is broadcasted. The same happens when connection is broken.
This is all done when `SatelModule.open` is called (SatelBridgeHandler, line 77). It may seem a bit complicated, because it is writen as a library with some level of abstraction. For example `SatelBridgeHandler` class is an adapter for `SatelModule`.

kaikreuzer replied:
Ok, then it was my mistake to not see this, sorry.
Please let me know, once you have addressed the other comments and we shouldn't be far away from the merge anymore.

druciak commented on addons/binding/org.openhab.binding.satel/ESH-INF/thing/output.xml:
```diff
@@ -0,0 +1,40 @@
+		<description>Output from the alarm system.</description>
+
+		<channels>
+			<channel typeId="state" id="state">
```
This is a binary output and the channel reflects its state. IMO "state" name describes channel purpose the best. Do you have better name for such channel?

kaikreuzer replied:
Maybe "output" as this is also what you use for the label? But my comment was more general as you are using the very same channel type at many other places and also have types like `state_adv` and `state_adv_ro`, which you re-use all over the place for functionally very different types of channels.

druciak replied:
Again, `state`, `state_ro` etc only defines type of channel, while ids, labels are overriden in each single channel. This is a kind of base definition so I am not sure what you mean.

kaikreuzer replied:
See the link that I posted above:

_The granularity of channel types should be on its semantic level, i.e. very fine-grained: If a Thing measures two temperature values, one for indoor and one for outdoor, this should be modeled as two different channel types. Overriding labels of a channel type must only be done if the very same functionality is offered multiple times, e.g. having an actuator with 5 relais, which each is a simple “switch”, but you want to individually name the channels (1-5)._

druciak replied:
My fault indeed. However I don't understand why is that. To clarify: I don't expect your explanation. :)

druciak replied:
Done.

druciak commented on addons/binding/org.openhab.binding.satel/src/main/java/org/openhab/binding/satel/handler/SatelSystemHandler.java:
```diff
@@ -0,0 +1,138 @@
+        Collection<SatelCommand> result = new LinkedList<>();
+        boolean anyStatusChannelLinked = getThing().getChannels().stream()
+                .filter(channel -> STATUS_CHANNELS.contains(channel.getUID().getId()))
+                .anyMatch(channel -> !linkRegistry.getLinks(channel.getUID()).isEmpty());
```
That's not true. `linkRegistry` is a protected field and that means it is supposed to be used by subclasses of `BaseThingHandler`. I would use `isLinked` but it lacks of version with channel object as a parameter and since I already have a channel this is a kind of optimization.

kaikreuzer replied:
See https://github.com/eclipse/smarthome/pull/5161

kaikreuzer replied:
> I would use isLinked but it lacks of version with channel object as a parameter

Well, you can easily take the id from your channel object and use this method with it.

druciak replied:
You're referencing something that is **hours** old in a PR 9 moths old. That's not fair...

kaikreuzer replied:
Haha, indeed :-)
I didn't blame you for not knowing that it wasn't supposed to be used. I only created the linked PR in order to be able to blame people in future if they use it 😎 .

druciak replied:
To be honest most parts of the code is over 2 years old and I am not talking about files ported from OH 1.x binding. Since API was changing frequently that time I decided to hold on with publishing 2.x version. 
It seems it still applies to the current OH2 status. In such case I will be chasing the current API in this PR and it will be never merged. I am just wondering how it goes with other bindings that has been merged recently.

kaikreuzer replied:
The API is pretty stable and we try our very best not to break anything. FTR, `linkRegistry` was NEVER to be used by anyone and I had an eye on this on every review. it is only now that I took action and put a deprecated marker on it, so that it is clear to everyone in the first place that it must not be used.

druciak replied:
Done.

kaikreuzer commented on addons/binding/org.openhab.binding.satel/ESH-INF/thing/partition.xml:
```diff
@@ -0,0 +1,104 @@
+				<description></description>
+			</channel>
+			<channel typeId="state_adv_ro" id="exit_time_gt_10">
+				<label>Exit time greater than 10 seconds</label>
```
"Exit > 10s"?

druciak commented on addons/binding/org.openhab.binding.satel/src/main/java/org/openhab/binding/satel/internal/SatelHandlerFactory.java:
```diff
@@ -0,0 +1,130 @@
+        discoveryServiceRegistrations.put(bridgeHandler.getThing().getUID(), discoveryServiceRegistration);
+    }
+
+    private ThingUID getDeviceUID(ThingTypeUID thingTypeUID, ThingUID thingUID, Configuration configuration,
```
This method is good for creating thing UID basing on device ID (instead of random generated).

druciak commented on addons/binding/org.openhab.binding.satel/src/main/java/org/openhab/binding/satel/internal/SatelHandlerFactory.java:
```diff
@@ -0,0 +1,130 @@
+    }
+
+    @Override
+    public Thing createThing(ThingTypeUID thingTypeUID, Configuration configuration, ThingUID thingUID,
```
To assign UID to a thing based on device's ID.

kaikreuzer replied:
Hm, ok, I never thought about doing this in a binding, but it sounds ok to do it.

druciak replied:
There are plenty of bindings doing this, I assume all of them were accepted (by you or someone else in charge).

kaikreuzer replied:
Really? First time I notice it at least. Just out of curiosity, do you have some examples of other bindings?

druciak replied:
Well maybe not "plenty". ;-) 
I saw this at least in two bindings, the one I recall is DSC alarm.

druciak replied:
I will remove it if you think it is not right.

kaikreuzer replied:
No, it is fine for me, you can leave it if you like.

druciak commented on addons/binding/org.openhab.binding.satel/ESH-INF/thing/output.xml:
```diff
@@ -0,0 +1,40 @@
+		<channels>
+			<channel typeId="state" id="state">
+				<label>Output</label>
+				<description></description>
```
Done.

druciak commented on addons/binding/org.openhab.binding.satel/src/main/java/org/openhab/binding/satel/internal/command/ControlObjectCommand.java:
```diff
@@ -0,0 +1,72 @@
+
+    private static final long REFRESH_DELAY = 1000;
+
+    private static final Timer REFRESH_TIMER = new Timer("ControlObjectCommand timer", true);
```
Done.

druciak commented on addons/binding/org.openhab.binding.satel/src/main/java/org/openhab/binding/satel/internal/command/IntegraVersionCommand.java:
```diff
@@ -0,0 +1,94 @@
+    protected boolean isResponseValid(SatelMessage response) {
+        // validate response
+        if (response.getCommand() != COMMAND_CODE) {
+            logger.error("Invalid response code: {}", response.getCommand());
```
Done.

druciak commented:
Hey @kaikreuzer, @martinvw 
I've just pushed requested changes. Please take a look if all is good.

kaikreuzer commented:
Thanks for your updates - let's go ahead and finally merge it!

druciak commented:
Thank you!
